### PR TITLE
Mon working in progress

### PR DESCRIPTION
**Pull request by yunstanford**
Graphite Sending Part.

For Code Review.

Not Test Yet.

### CODE REVIEW

File: aiographite/aiographitesend.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class AsyncioGraphiteSendService(object):
+
+	def __init__(self, graphite_server, graphite_port, protocol = "pickle"):
+		"""
+			Must call __init__ before use
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I think you probably don't need this comment. it's impossible to call a method without instantiating the object first.

File: aiographite/aiographitesend.py [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+		"""
+		# Graphite Server Address
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you probably don't need this comment, since the code is self-explanatory.

I'm not against commenting code: I think it's great. But I also think that readable code ensures we don't need comments, a lot of the time.

File: aiographite/graphite_escaping.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	"""
+	display_metric_name = ""
+	try:
+		display_metric_name = urllib.unquote(idna_str).decode('idna').encode('utf-8')
+	except Exception, e:
+		raise e
+	return display_metric_name
+
+
+def dummy_data():
```
Comment (mayurmahajan) [inline aiographite/graphite_escaping.py]: cool examples. maybe later you want to remove this funtion and use these examples as part of your unit tests to check encoding/decoding functionality.

File: aiographite/graphite_escaping.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		raise e
+	return valid_graphite_metric_name
+	
+
+def metrics_name_from_graphite(idna_str):
+	"""
+		@param: valid metric name of graphite
+		@return: the original name
+	"""
+	display_metric_name = ""
```
Comment (mayurmahajan) [inline aiographite/graphite_escaping.py]: do you need this initialization?

File: aiographite/graphite_escaping.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	very long string: exceed the 63-character
+
+"""
+
+
+"""
+	For special characters :
+	1. 	The dot (.) is a special character because it delineates each metric’s path component, 
+		but this is an easy fix; just substitute all dots for underscores or '%2E'. 
+		For example, www.zillow.com => www_zillow_com
```
Comment (mayurmahajan) [inline aiographite/graphite_escaping.py]: if you chose to substitute . with %2E then I would suggest using %2E in example and removing "_". as it creates bit confusion.

File: aiographite/aiographitesend.py [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self.socket = socket.create_connection(self.graphite_server_address)
+			self.socket.setblocking(False)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self.graphite_server_address)
+		except Exception as e:
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you don't need to check this exception. Python will raise if there isn't an exception is recognizes.

File: aiographite/aiographitesend.py [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+
+
+	def send_single_valid_data(self, metric, value, timestamp = None):
+		"""
+			@metric: String
+			@value: int
+			@timestamp: int
+			Send a single data(metric value timestamp) to graphite
+		"""
+		timestamp = int(time.time()) if timestamp is None else int(timestamp)
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: this could be:

`int(timestamp or time.time())`

File: aiographite/aiographitesend.py [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+		else:
+			valid_dataset = [(self.to_graphite_valid_metric_name(metric_dir_list), value, timestamp) for metric_dir_list, value, timestamp in dataset]
+
+		self.send_valid_dataset_list(valid_dataset, timestamp)
+
+
+
+	def send_single_valid_data(self, metric, value, timestamp = None):
+		"""
+			@metric: String
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: Documentation on types is great. In python 3.5 and above, you can use the typing module:

https://docs.python.org/3/library/typing.html

python3 supports type annotations, so we should use those.

We can then support mypy, the preferred python tool for static type checking: http://mypy-lang.org/

Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: for Python 2, we should use sphinx's documentation syntax:

http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists

File: aiographite/aiographitesend.py [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+DEFAULT_GRAPHITE_PLAINTEXT_PORT = 2003
+DEFAULT_GRAPHITE_PICKLE_PORT = 2004
+SUPPORT_PROTOCOLS = ["plaintext", "pickle"]
+
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class AsyncioGraphiteSendService(object):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I'd probably name this "AIOGraphiteSendService" or something, instead of AsyncIOGraphite. I know it doesn't match up with the asyncio name, but everything else in this module (egg name, exception name, etc) uses "AIO"

Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: Also, we don't really need "SendService". 
- the word 'service' is unnecessary, it doesn't describe the class any better.
- the word 'send' is unnecessary. It's clear from context that this class can send:

```
graphite = AsyncIOGraphite()
graphite.send_single_data()
```

Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I would probably just call this "AIOGraphite". It's very descriptive, even with that subset.

File: aiographite/aiographitesend.py [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+
+	def __init__(self, graphite_server, graphite_port, protocol = "pickle"):
+
+		self.graphite_server_address = (graphite_server, graphite_port)
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: private / protected variables should have a leading undescore:

```
self._graphite_server_address
```

this goes for both fields and methods:

```
self._connect_to_graphite()
```

It helps future developers understand what they can and can't modify. If everything is "public" developers don't really know what they could refactor without breaking others, and consumers assume they can use everything.

File: aiographite/aiographitesend.py [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+
+		self.connect_to_graphite()
+
+		self.protocol = protocol
+
+		self.loop = asyncio.get_event_loop()
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: for the event loop, it would be nice allow someone to pass the event loop as well, kind of like aiohttp:

http://aiohttp.readthedocs.io/en/stable/#getting-started

We can default to asyncio.get_event_loop() though, if one is not passed.

File: aiographite/aiographitesend.py [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+			else:
+				(metric, value, data_timestamp) = data
+				timestamp = data_timestamp
+			listOfData.append(formate_function(metric, value, timestamp))
+		message =  generate_message_function(listofData)
+		return message
+
+
+
+	def generate_message_for_plaintext(self, listOfPlaintext):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I see that for plaintext and pickle, there's some custom commands we have to run for those:

```
generate_message
formatted_data
```

If I see a lot of behavior that depends on an enum value, we can convert those to classes, and pass in classes instead:

```
graphite_plaintext = AIOGraphite(protocol=PlaintextProtocol())
graphite_pickle = AIOGraphite(protocol=PickleProtocol())
```

this is also very cool, because it allows other developers to pass in their own protocol as well.

File: aiographite/graphite_escaping.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	print(to_graphite)
+	print('=== Convert the Graphite valid metric name back to the original name >>>')
+	from_graphite = metrics_name_from_graphite(to_graphite)
+	print(from_graphite)
+	print('============= Test Consistency ===============', name == from_graphite)
+
+def main():
+	dummy_data()
+
+if __name__ == '__main__':
```
Comment (toumorokoshi) [inline aiographite/graphite_escaping.py]: yeah, we should remove the **name** stuff to. We have pytest for unit testing. running the individual modules themselves for testing is an interesting idea, but leads to a lot of boilerplate code that distracts users from the real code.

File: tests/test_consistency.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    'abc @edf#',
+    'abc.@edf#',
+    'abc_ @ e_df#',
+    'a.b.c_ @ e_df#',
+    'a.b.___c d _feg',
+    '_ . .fda',
+    '_.',
+    '汉 字.汉*字'
+])
+def test_consistency(name):
```
Comment (toumorokoshi) [inline tests/test_consistency.py:19]: cool. Great tests!

File: aiographite/graphite_escaping.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+		special characters to make it valid for Graphite, and then URL decode it when we need to
+		reconstruct the information
+"""
+
+
+
+
+# Naming Metrics:
+# <section_name>.<section_name>.<section_name>.<section_name>
+def metrics_name_to_graphite(section_name):
```
Comment (toumorokoshi) [inline aiographite/graphite_escaping.py]: this method could be more descriptive. It's not sending metric information to graphite.

I would say "encode_metric_name_for_graphite" and "decode_metric_name_for_graphite"

Or, what might be nicer is a static class:

```
metric_name = GraphiteEncoder.encode(section_name)
```

File: aiographite/graphite_escaping.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+	So any special name not allow to appear in directory/file name is not supported by Graphite.
+"""
+
+"""
+	Naming metrics schema:
+		<namespace>.<instrumented section>.<target (noun)>.<action (past tense verb)>
+"""
+
+"""
+	Limitation and explaination:
```
Comment (toumorokoshi) [inline aiographite/graphite_escaping.py]: is this copy pasted from the documentation? we should probably just link, in that case.

copy-paste can become out of date.

File: aiographite/aiographitesend.py [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+
+
+
+#########################################################
+#########################################################
+#########################################################
+
+
+# Module Instance Variable
+aiographite_send_instance = None
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I don't think we need those methods.

If someone wants to make a global aiographite_send, they can do that themselves.

In general, these types of globals are dangerous, because they can lead to race conditions and assumptions about configuration that wasn't correct.

File: aiographite/aiographitesend.py [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+			self.socket.shutdown(1)
+		except AttributeError:
+			self.socket = None
+		except Exception:
+			self.socket = None
+		finally:
+			self.socket = None
+
+
+	def close_event_loop(self):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: do we need this call? I would hope consumers would know to close the loop directly, or use asyncio.get_event_loop.close().

File: aiographite/aiographitesend.py [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+			message = self.generate_message_for_data_list(dataset, timestamp, self.plaintext_protocol_formatted_data, self.generate_message_for_plaintext)
+		else:
+			message = self.generate_message_for_data_list(dataset, timestamp, self.pickle_protocol_formatted_data, self.generate_message_for_pickle)
+
+		# Sending Data
+		self.loop.run_until_complete(asyncio.ensure_future(self.send_message(message)))
+
+
+
+	def send_valid_dataset_dict(self, dataset, timestamp = None):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: we don't need this.... I think a list is sufficient.

File: aiographite/aiographitesend.py [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+
+			If you're very confident that the metric name is valid, then use <method: send_single_valid_data> instead.
+
+		"""		
+		valid_metric_name = self.to_graphite_valid_metric_name(metric_dir_list)
+		self.send_single_valid_data(valid_metric_name, value, timestamp)
+
+
+
+	def send_dataset_list(self, dataset, timestamp = None):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I see a lot of combinational methods here. Specifically combinations of:
1. dataset submissions (list vs single)
2. conversions from the various inputs.

If you try to support all permutations of this, it's an N^2 problem. Instead, we should focus on a bunch of smaller, composable methods that we can string together as necessary.

e.g.

```
metric_names = [self.to_graphite_metric_list(metric_parts) for metric_parts in metric_list]
self.send_to_graphite(metric_names)
```

vs 

```
self.send_metric_names_to_graphite(metric_list)
```

File: aiographite/aiographitesend.py [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+			Send a single data(metric value timestamp) to graphite
+		"""
+		timestamp = int(time.time()) if timestamp is None else int(timestamp)
+		message = ""
+		if self.protocol == "plaintext":
+			message = plaintext_protocol_formatted_data(metric, value, timestamp)
+		else:
+			listOfMetricTuples = [pickle_protocol_formatted_data(metric, value, timestamp)]
+			message = generate_message_for_pickle(listOfMetricTuples)
+		self.loop.run_until_complete(asyncio.ensure_future(self.send_message(message)))
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: these functions should be asychronous. Using the event loop's run_until_complete makes this call synchronous.

We'll be using these functions in our aiohttp service, which uses coroutines:

https://docs.python.org/3/library/asyncio-task.html#example-hello-world-coroutine

All of these calls should also be coroutines.

Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: It looks like sockets are synchronous, so they won't work really well here. 

Let's use asyncio's connections instead: that allows for asynchronous IO:
http://www.curiousefficiency.org/posts/2015/07/asyncio-tcp-echo-server.html

File: aiographite/aiographitesend.py [hunk 22]
```diff
@@ -0,0 +1,325 @@
...
+DEFAULT_GRAPHITE_PLAINTEXT_PORT = 2003
+DEFAULT_GRAPHITE_PICKLE_PORT = 2004
+SUPPORT_PROTOCOLS = ["plaintext", "pickle"]
+
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class PlaintextProtocol(object):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: this file is still a bit long: ~ 300 lines. We should be shooting for ~ 100 lines, if possible.

We should move the protocols to a different module. That helps keep small context on your code.

```
from .protocol import PlaintextProtocol, PickleProtocol
# or
from .protocol.plaintext import PlaintextProtocol
from .protocol.pickle import PickleProtocol
```

File: aiographite/aiographitesend.py [hunk 23]
```diff
@@ -0,0 +1,325 @@
...
+			return the required message formate for protocol 'plaintext'
+			@param: 
+				listOfPlaintext: ["metric1 value1 timestamp1", "metric2 value2 timestamp2", ...]
+				type: List of String
+		"""
+		return "".join(listOfPlaintext).encode('ascii')
+		
+
+
+class PickleProtocol(object):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you don't need to inherit from object in Python 3:

http://stackoverflow.com/questions/54867/what-is-the-difference-between-old-style-and-new-style-classes-in-python

```
class PickleProtocol:
```

File: aiographite/aiographitesend.py [hunk 24]
```diff
@@ -0,0 +1,325 @@
...
+
+	def data_formate_function(self, metric: str, value: int, timestamp: int):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: typo: "data_format_function" (no e)

File: aiographite/aiographitesend.py [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+		"""
+			@return: required data formate when sending data through 'pickle' protocol
+			@return_type: Tuple
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you can use python 3 typing here:

https://docs.python.org/3/library/typing.html#typing.Tuple

```
timestamp: int) -> Tuple[str, Tuple[int, int]]:
```

File: aiographite/aiographitesend.py [hunk 26]
```diff
@@ -0,0 +1,325 @@
...
+
+
+
+
+class AIOGraphite(object):
+
+	def __init__(self, graphite_server, graphite_port = DEFAULT_GRAPHITE_PICKLE_PORT, protocol = PickleProtocol(), loop = None):
+
+		self._graphite_server = graphite_server
+
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py:26]: I would probably remove these spaces. They don't really add to the readability, and add extra lines that makes it hard to read more code on the same page.

File: aiographite/aiographitesend.py [hunk 27]
```diff
@@ -0,0 +1,325 @@
...
+		header = struct.pack("!L", len(payload))
+		message = header + payload
+		return message
+
+
+
+
+class AIOGraphite(object):
+
+	def __init__(self, graphite_server, graphite_port = DEFAULT_GRAPHITE_PICKLE_PORT, protocol = PickleProtocol(), loop = None):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you want to avoid object instantiation as default arguments. The class does not regenerate those arguments every time it's called.

try this, for an example:

```
def foo(x={}, k, v):
    x[k]= v
    print(foo)

foo("a", "b") # => {"a": "b"}
foo("c", "d") # => {"a: "b", "c": "d"}
```

File: aiographite/aiographitesend.py [hunk 28]
```diff
@@ -0,0 +1,325 @@
...
+	def data_formate_function(self, metric: str, value: int, timestamp: int):
+		"""
+			@return: required data formate when sending data through 'plaintext' protocol
+			@return_type: String
+		"""
+		formatted_data = " ".join([metric, str(value), str(timestamp)])
+		return formatted_data + "\n"
+
+
+	def generate_message_function(self, listOfPlaintext: List[str]):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: I don't think we need "function" in the function name: we know it's a function already, by reading documentation.

File: aiographite/aiographitesend.py [hunk 29]
```diff
@@ -0,0 +1,325 @@
...
+
+
+	@asyncio.coroutine
+	async def disconnect(self):
+		"""
+			Close the TCP connection 
+		"""
+		try:
+			self._writer.close()
+		except AttributeError:
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: you can catch multiple exceptions using a tuple:

```
except (AttributeError, Exception):
```

But in this case, you already have the 'finally' block. So you don't need the others.

Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: good call on the finally, btw. Great habit to start this early 👍

File: aiographite/aiographitesend.py [hunk 30]
```diff
@@ -0,0 +1,325 @@
...
+	@asyncio.coroutine
+	def connect_to_graphite(self):
+		"""
+			Connect to Graphite Server based on Provided Server Address
+		"""
+		try:
+			self._reader, self._writer = yield from asyncio.open_connection(self._graphite_server, self._graphite_port, loop = self.loop)
+		except socket.gaierror:
+			raise AioGraphiteSendException("Unable to connect to the provided server address %s:%s" % self._graphite_server_address)
+		except Exception as e:
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: no need for this. exception will be raised automatically

File: aiographite/aiographitesend.py [hunk 31]
```diff
@@ -0,0 +1,325 @@
...
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of String
+		"""
+		return ".".join([GraphiteEncoder.encode(dir_name) for dir_name in metric_dir_list])
+
+
+
+	@property
+	def reader(self):
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: do we want to expose the reader / writer publicly? I think we shouldn't, unless there's a good reason to.

File: aiographite/aiographitesend.py [hunk 32]
```diff
@@ -0,0 +1,325 @@
...
+	
+
+
+#########################################################
+#########################################################
+#########################################################
+
+
+
+def _dummy_message_plaintext_formate():
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: move these to unit tests, or maybe the Protocol class?

File: aiographite/aiographitesend.py [hunk 33]
```diff
@@ -0,0 +1,325 @@
...
+	print("[(path1, (timestamp1, value1)), (path2, (timestamp2, value2)), ...]")
+
+
+
+def main():
+	_dummy_message_plaintext_formate()
+
+
+
+if __name__ == '__main__':
```
Comment (toumorokoshi) [inline aiographite/aiographitesend.py]: no main functions, or making the file callable. We can unit test using pytest.

File: aiographite/graphite_escaping.py [hunk 34]
```diff
@@ -0,0 +1,106 @@
...
+		but this is an easy fix; just substitute all dots for underscores or '%2E'. 
+		For example, www.zillow.com => www_zillow_com / www%2Ezillow%2Ecom
+	2   For the rest of the special characters(except dot), just URL any metric name with 
+		special characters to make it valid for Graphite, and then URL decode it when we need to
+		reconstruct the information
+"""
+
+
+
+class GraphiteEncoder:
```
Comment (toumorokoshi) [inline aiographite/graphite_escaping.py]: cool. love it.

File: tests/test_graphite_send.py [hunk 35]
```diff
@@ -0,0 +1,128 @@
...
+
+def teardown_module(module):
+	"""
+		1. Destroy aiographite_send instance
+		2. Close the server socket
+	"""
+	try:
+		module.server.shutdown(socket.SHUT_RD)
+		module.server.close()
+	except Exception:
```
Comment (toumorokoshi) [inline tests/test_graphite_send.py]: we probably want this to raise an exception. I'd worry about the cleanup not working properly here.

File: aiographite/aiographitesend.py [hunk 36]
```diff
@@ -0,0 +1,325 @@
...
+
+class AioGraphiteSendException(Exception):
+    pass
+
+
+class PlaintextProtocol(object):
+
+	def data_formate_function(self, metric: str, value: int, timestamp: int):
+		"""
+			@return: required data formate when sending data through 'plaintext' protocol
```
Comment (mayurmahajan) [inline aiographite/aiographitesend.py]: typo: format

File: tests/test_graphite_send.py [hunk 37]
```diff
@@ -0,0 +1,128 @@
...
+from asyncio import test_utils
+
+
+
+DEFAULT_GRAPHITE_PLAINTEXT_PORT = 2003
+DEFAULT_GRAPHITE_PICKLE_PORT = 2004
+
+
+
+def setup_module(module):
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py]: can you add docstring for this method.

File: tests/test_graphite_send.py [hunk 38]
```diff
@@ -0,0 +1,128 @@
...
+		module.server.shutdown(socket.SHUT_RD)
+		module.server.close()
+	except Exception:
+		pass
+	module.server = None
+
+
+
+@pytest.fixture
+def metric_dir_list():
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py]: are these fixtures for later use? you need to put them in conftest.py file and then you can use them in your unit tests .

File: tests/test_graphite_send.py [hunk 39]
```diff
@@ -0,0 +1,128 @@
...
+	assert pickle.generate_message_function(tuple_list) == expected_message
+
+
+
+def test_generate_message_for_plaintext():
+	plaintext_list = ["metric1 value1 timestamp1\n", "metric2 value2 timestamp2\n", "metric3 value3 timestamp3\n"]
+	expected_message = b'metric1 value1 timestamp1\nmetric2 value2 timestamp2\nmetric3 value3 timestamp3\n'
+	plaintext = PlaintextProtocol()
+	message = plaintext.generate_message_function(plaintext_list)
+	assert message == expected_message
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py:47]: awesome!!!

File: tests/test_graphite_send.py [hunk 40]
```diff
@@ -0,0 +1,128 @@
...
+	plaintext = PlaintextProtocol()
+	assert plaintext.data_formate_function(metric, value, timestamp) == " ".join([metric, str(value), str(timestamp)]) + "\n"
+
+
+
+def test_generate_message_for_pickle():
+	pickle = PickleProtocol()
+	tuple_list = [('a', (123, 456)), ('b', (123, 456)), ('c', (876, 987))]
+	expected_message = b'\x00\x00\x00B\x80\x02]q\x00(X\x01\x00\x00\x00aq\x01K{M\xc8\x01\x86q\x02\x86q\x03X\x01\x00\x00\x00bq\x04K{M\xc8\x01\x86q\x05\x86q\x06X\x01\x00\x00\x00cq\x07Ml\x03M\xdb\x03\x86q\x08\x86q\te.'
+	assert pickle.generate_message_function(tuple_list) == expected_message
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py]: these tests are so easy to understand. nice!!

File: tests/test_graphite_send.py [hunk 41]
```diff
@@ -0,0 +1,128 @@
...
+
+
+
+def test_generate_message_for_plaintext():
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py:42]: this test is checking for happy case, what happens when the list is null/list of empty strings etc

File: tests/test_graphite_send.py [hunk 42]
```diff
@@ -0,0 +1,128 @@
...
+    ('hotpad', 53534, 32425),
+])
+def test_pickle_protocol_formatted_data(metric, value, timestamp):
+	pickle = PickleProtocol()
+	assert pickle.data_formate_function(metric, value, timestamp) == (metric, (timestamp, value))
+
+
+
+@pytest.mark.parametrize("metric, value, timestamp", [
+    ('M M M', 124, 234),
```
Comment (mayurmahajan) [inline tests/test_graphite_send.py]: can add a test where metric field is null.

File: aiographite/aiographitesend.py [hunk 43]
```diff
@@ -0,0 +1,325 @@
...
+		"""
+		if not dataset:
+			return 
+
+		if len(dataset[0]) == 2:
+			valid_dataset = [(self._to_graphite_valid_metric_name(metric_dir_list), value) for metric_dir_list, value in dataset]
+		else:
+			valid_dataset = [(self._to_graphite_valid_metric_name(metric_dir_list), value, timestamp) for metric_dir_list, value, timestamp in dataset]
+
+		self.send_valid_metric_list(valid_dataset, timestamp)
```
Comment (mayurmahajan) [inline aiographite/aiographitesend.py]: need an await here?

File: aiographite/aiographitesend.py [hunk 44]
```diff
@@ -0,0 +1,325 @@
...
+					metric_dir_list = [metaccounts, authentication, password, attempted]
+
+			@metric_dir_list: List of string
+			@timestamp: the type should be int
+
+			If you're very confident that the metric name is valid, then use <method: send_single_valid_data> instead.
+
+		"""		
+		valid_metric_name = self._to_graphite_valid_metric_name(metric_dir_list)
+		self.send_single_valid_metric(valid_metric_name, value, timestamp)
```
Comment (mayurmahajan) [inline aiographite/aiographitesend.py]: need an await here?
